sirepo/auth_db/__init__.py

```diff
@@ -0,0 +1,294 @@
+
+
+def all_uids(qcall):
+    return UserRegistration.search_all_for_column("uid")
+
+
+def audit_proprietary_lib_files(qcall, force=False, sim_types=None):
+    """Add/removes proprietary files based on a user's roles
+
+    For example, add the Flash tarball if user has the flash role.
+
+    Args:
+      qcall (quest.API): logged in user
+      force (bool): Overwrite existing lib files with the same name as new ones
+      sim_types (set): Set of sim_types to audit (proprietary_sim_types if None)
+    """
+    from sirepo import sim_data, simulation_db
+
+    def _add(proprietary_code_dir, sim_type, sim_data_class):
+        p = proprietary_code_dir.join(sim_data_class.proprietary_code_tarball())
+        with simulation_db.tmp_dir(chdir=True, qcall=qcall) as t:
+            d = t.join(p.basename)
+            d.mksymlinkto(p, absolute=False)
+            subprocess.check_output(
+                [
+                    "tar",
+                    "--extract",
+                    "--gunzip",
+                    f"--file={d}",
+                ],
+                stderr=subprocess.STDOUT,
+            )
+            # lib_dir may not exist: git.radiasoft.org/ops/issues/645
+            l = pykern.pkio.mkdir_parent(
+                simulation_db.simulation_lib_dir(sim_type, qcall=qcall),
+            )
+            e = [f.basename for f in pykern.pkio.sorted_glob(l.join("*"))]
+            for f in sim_data_class.proprietary_code_lib_file_basenames():
+                if force or f not in e:
+                    t.join(f).rename(l.join(f))
+
+    s = sirepo.feature_config.proprietary_sim_types()
+    if sim_types:
+        assert sim_types.issubset(
+            s
+        ), f"sim_types={sim_types} not a subset of proprietary_sim_types={s}"
+        s = sim_types
+    u = qcall.auth.logged_in_user()
+    for t in s:
+        c = sim_data.get_class(t)
+        if not c.proprietary_code_tarball():
+            continue
+        d = sirepo.srdb.proprietary_code_dir(t)
+        assert d.exists(), f"{d} proprietary_code_dir must exist" + (
+            "; run: sirepo setup_dev" if pykern.pkconfig.channel_in("dev") else ""
+        )
+        r = UserRole.has_role(
+            qcall=qcall,
+            role=sirepo.auth_role.for_sim_type(t),
+        )
+        if r:
+            _add(d, t, c)
+            continue
+        # SECURITY: User no longer has access so remove all artifacts
+        pykern.pkio.unchecked_remove(simulation_db.simulation_dir(t, qcall=qcall))
+
+
+def create_or_upgrade(qcall):
+    from sirepo import db_upgrade
+
+    UserDbBase.metadata.create_all(bind=_engine)
+    db_upgrade.do_all(qcall)
+
+
+def db_filename():
+    return sirepo.srdb.root().join(_SQLITE3_BASENAME)
+
+
+def init_module():
+    def _classes():
+        res = PKDict()
+        for r in sirepo.feature_config.cfg().package_path:
+            p = pykern.pkinspect.module_name_join((r, "auth_db"))
```

Comment:
This isn't backwards compatible. Packages on package path may not provide an auth_db sub-package.

Reply:
I was wondering whether to use package_path or not. I'm happy to remove. The failure is immediate so not a huge problem. Should I remove?

Reply:
What about adding a check? If the package has it then use it. That is what the other areas of the code do in regards to package path.

Reply:
This is a trickier case, because the other cases are looking for a module along package_path. This is loading all modules along package_path. #3933 is key here. If there's a syntax error in `auth_db/__init__.py`, it gets a not found, and there won't be an sirepo.auth_db module. The other issue is if the package path is being used to improve auth_db model or creating a new one. We don't handle the former case, and I don't think we can support the latter case. Therefore, I think this is a YAGNI, and I'll take it out.

Reply:
Sounds good.